Wrap Ops during etuplization

rlouf opened the pull request:
Evaluating etuplized objects fails for some `RandomVariable` ops whose
`__call__` function does not defer to `make_node`.

In this commit we wrap Ops during etuplization with a class that always
defers `__call__` to `make_node`. We also add a dispatch rule for
`etuplize` so it also wraps `RandomVariable` with the same class.

Closes #1002.


**Thank you for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [ ] There are tests covering the changes introduced in the PR.

rlouf commented:
Some tests that are unrelated to `etuplize`ation expectedly fail. They're an easy fix, but still it makes we wonder whether we shouldn't come up with something in `etuples` so that this fix only requires the following code:

```python
@dataclass
class MakeRVNodeOp:
    op: Op

    def __call__(self, *args):
        return self.op.make_node(*args)

@etuplize.register(RandomVariable)
def etuplize_random(*args, **kwargs):
    return etuple(MakeRVNodeOp, etuplize.funcs[(object,)](*args, **kwargs))
```

rlouf commented:
Codecov complains about a file that is not affected by the changes. @brandonwillard this is ready for review.

rlouf commented:
We /should/ be able to unify `etuple(op, ...)` with the result of `etuplize(op(...))`, so I don't think this solution (wrapping the Ops) is acceptable.

I gave it more thought, and I think that the solution that introduces the least amount of change to `Aesara` while not being too hacky is to have a function handle [the object's evaluation in `etuples`](https://github.com/pythological/etuples/blob/1ed56e5226051244aac4ef64feb9615af0d5bd28/etuples/core.py#L204) and dispatch this function depending on the type of `op`.

In the present case we would only need to add a function in`aesara.graph.rewriting.unify` that would be along the lines of:

``` python
def eval_Op(op, evaled_args, evaled_kwargs):
    return op.make_node(*evaled_args, **evaled_kwargs)
    
_eval.add((Op,), eval_Op)
```

In fact, couldn't we used `apply` for that? [It is already dispatched for  `Op`](https://github.com/aesara-devs/aesara/blob/9f176da71d41635e8854fd601fd6a68102b0c6e5/aesara/graph/rewriting/unify.py#L139).

brandonwillard commented:
> I gave it more thought, and I think that the solution that introduces the least amount of change to `Aesara` while not being too hacky is to have a function handle [the object's evaluation in `etuples`](https://github.com/pythological/etuples/blob/1ed56e5226051244aac4ef64feb9615af0d5bd28/etuples/core.py#L204) and dispatch this function depending on the type of `op`.

One of the first things I considered was a change to `ExpressionTuple._eval_step` that would allow one to specify something other than the use of `__call__` (e.g. `op(*args)`) for evaluation; however, that's a very critical point in the `ExpressionTuple` evaluation process, and we don't need/want dispatching there&mdash;at least if it's not necessary.

I left that approach with the idea that we could simply subclass `ExpressionTuple` and make that change.  We could expose an interface for that exact kind of customization, though, which would make such subclasses very simple.  For instance, adding a `ExpressionTuple._eval_apply` that takes `op` and `op_args` values would help.

brandonwillard commented:
https://github.com/pythological/etuples/pull/19 illustrates the `ExpressionTuple` changes.  We would need to follow that up here by creating a custom `ExpressionTuple` subclass that uses `Op.make_node` in its `ExpressionTuple._eval_apply`.

brandonwillard reviewed (commented):
Looks good.  The only thing I would consider changing is the limited use of a custom `ExpressionTuple`.  We should _always_ use `Op.make_node`.

rlouf commented:
I considered that, it's just that the apply node created by  `RandomVariable`s have the rng as well as the output variable as outputs, and we're only interested in returning the latter. I could add an `OpExpressionTuple` for other Ops.

brandonwillard commented:
> I considered that, it's just that the apply node created by `RandomVariable`s have the rng as well as the output variable as outputs, and we're only interested in returning the latter. I could add an `OpExpressionTuple` for other Ops.

We're always interested in all outputs of an `Op`.  We might not be using them in our current rewrites, but they need to be available at some point.  Regardless, we really shouldn't be "special casing" `RandomVariable`s in this way.

rlouf commented:
> Regardless, we really shouldn't be "special casing" `RandomVariable`s in this way.

I'm not sure about this one. Remember that this issue was created to address an issue with etuplization of `RandomVariable`s _specifically_.

rlouf commented:
Evaluating `Op`s now returns an `Apply` node, as opposed to a `TensorVariable` (or an error in the case of `RandomVariable`s) on the `main` branch. Is that what we want? Asking making changes elsewhere in the codebase.

brandonwillard commented:
> Evaluating `Op`s now returns an `Apply` node, as opposed to a `TensorVariable` (or an error in the case of `RandomVariable`s) on the `main` branch. Is that what we want? Asking before making changes elsewhere in the codebase.

For ease of use with `kanren`, we might want to return the outputs list.

rlouf commented:
It is trickier than I thought because there is an assumption throughout the code that inputs to `Apply` nodes are passed as arguments individually. This causes issues when evaluating nested `OpExpressionTuple`s as `Apply` nodes interpret `[Var]` as `MakeVector(Var)`:

```python
import aesara
import aesara.tensor as at
from etuples import etuplize, etuple

a = at.scalar('a')
b = at.scalar('b')
c = at.scalar('c')

x_et = etuple(at.mul, etuple(at.add, b, b), etuple(at.add, a, a))
x_val = x_et.evaled_obj
print(x_val)
# [Elemwise{mul,no_inplace}.0] expected

print(x_val[0].owner.inputs)
# [MakeVector{dtype='float64'}.0, MakeVector{dtype='float64'}.0]

print(aesara.dprint(x_val[0].owner.inputs))
# MakeVector{dtype='float64'} [id A]
# |Elemwise{add,no_inplace} [id B]
#   |b [id C]
#   |b [id C]
# MakeVector{dtype='float64'} [id D]
```

Unless I misunderstood something, it is also assumed in `PatternNodeRewritter` that the evaluation of an `ExpressionTuple` will return a single variable, which was wrong even before this PR (I think?). I think this is touching on a much more general problem with regards to `Op`s with multiple outputs.

rlouf commented:
_Following up on a discussion with @brandonwillard_

This indeed touches on a much deeper issue, that is the fact that Aesara's IR lacks the ability to represent selections of specific multi-outputs outputs. All the logic is hidden in `Op.__call__` as the graph is being built.

To illustrate the problem, let's look again at the etuplization of `RandomVariable`s with the current state of the PR:

```python
#+begin_src python :results output
import aesara.tensor as at
from etuples import etuple, etuplize

u_rv = at.random.normal(0, 1)
u_et = etuplize(u_rv)
print(u_et)
# : oExpressionTuple((ExpressionTuple((<class 'aesara.tensor.random.ba...ormalRV'>, 'normal', 0, (0, 0), 'floatX', False)), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7FA6605E5EE0>), TensorConstant{[]}, TensorConstant{11}, TensorConstant{0}, TensorConstant{1}))

# reset the saved evaled obj
u_et._evaled_obj = u_et.null
print(u_et.evaled_obj)
# [normal_rv{0, (0, 0), floatX, False}.0, normal_rv{0, (0, 0), floatX, False}.out]
```

Of course we can hack our way around this, "special casing" the random variables as was mentioning above in the thread. It's unsatisfactory, but there's worse. Let's look at what happens when we etuplize one of the outputs of a multi-output `Op` and evaluate the resulting object:

```python
import aesara.tensor as at
from etuples import etuple, etuplize

a = at.matrix('a')
u, v, w = at.nlinalg.svd(a)
u_et = etuplize(u)
print(u_et)
# oExpressionTuple((ExpressionTuple((<class 'aesara.tensor.nlinalg.SVD'>, 1, 1)), a))

# reset the saved evaled obj
u_et._evaled_obj = u_et.null
print(u_et.evaled_obj)
# [SVD{full_matrices=1, compute_uv=1}.0, SVD{full_matrices=1, compute_uv=1}.1, SVD{full_matrices=1, compute_uv=1}.2]
```
We have lost some precious information in the process of etuplization, which is that `u` is the first output of the apply node created by `at.nlinalg.svd`, and are thus unable to evaluate the expression tuple to this variable. What we need to solve this issue is a more principled approach, and we can already have an idea about what this would look like in the etuplized version of the graph.

What we essentially need is adding an indexing operator when we etuplize the Ops, so that etuplizing the `RandomVariable` above would return

```python
ExpressionTuple(
    nth,
    1,
    oExpressionTuple((ExpressionTuple((<class 'aesara.tensor.random.ba...ormalRV'>, 'normal', 0, (0, 0), 'floatX', False)), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7FA6605E5EE0>), TensorConstant{[]}, TensorConstant{11}, TensorConstant{0}, TensorConstant{1}))
)
````

where `oExpressionTuple(...)` evaluates to the list of outputs of the Apply nodes. the `nth` function returns the element at index 1 in this list so the output is the `RandomVariable` returned by the `Apply` node.

This also works in the `at.nlinalg.svd` example:

```python
ExpressionTuple(
    nth,
    0,
    oExpressionTuple((ExpressionTuple((<class 'aesara.tensor.nlinalg.SVD'>, 1, 1)), a))
)
``` 

which evaluates to the first `Variable` in the outputs list of the `Apply` node. We may be able to get by temporarily using the information in the `Op`'s `default_output` attribute, and the `index` attribute of each `Variable`, but this doesn't solve the root issue that is the inability to represent the selection of outputs in Aesara's IR.

brandonwillard commented:
> This indeed touches on a much deeper issue, that is the fact that Aesara's IR lacks the ability to represent selections of specific multi-outputs outputs. All the logic is hidden in `Op.__call__` as the graph is being built.

N.B. This is also why a translation to something like e-graphs isn't entirely straightforward.

rlouf commented:
First I ended up creating a `nth` function that takes an index `n`, and apply node `node` and returns `node.outputs[n]`. I also updated the tests related to etuplization/evaluation because they were not really testing the evaluation (the evaluated object saved during etuplization instead). I added a test for multiple output nodes and nodes that have a default output. Both pass thanks to that `nth` function.

I believe this `nth` function should be promoted to an Op in Aesara's IR, so that `Op`s `__call__` function becomes:

```python
def __call__(self, *inputs, **kwargs):
    node = self.make_node(*inputs, **kwargs)
    outputs = node.outputs
   
    if len(outputs) == 0:
         return nth(0, outputs)
    elif:
        if self.default_output is not None:
            return nth(self.default_output, outputs)
    else:
         return [nth(i, output) for i, output in enumerate(outputs)]
```

We could probably rename `default_output` to `selected_output` and set its default to 0. The difficulty with this approach is that it changes the variables' `owner`, and a LOT of logic depends on the direct link `variable - apply node`. But it surely is worth it long-term.

Regarding multiple output `Op`s, I added support for the case where the caller calls `etuplize` of the list that contains both outputs by dispatching `car` to `cons` and `cdr` to `tuple(x) + ([],)`. I added a test to make sure it evaluates back to the list.

I now need to add tests around unification for multiple outputs and default output node. I am also not sure yet to what extent this is going to affect the rewrites.

rlouf commented:
Well, what I was discussing above has direct consequences today because of `default_output`:

```python
class MyDefaultOutputOp(Op):
    default_output = 1

    def make_node(self, *inputs):
        outputs = [MyType()(), MyType()()]
        return Apply(self, list(inputs), outputs)

    def perform(self, node, inputs, outputs):
        outputs[0] = np.array(np.array(inputs[0]))
        outputs[1] = np.array(np.array(inputs[1]))

x_at = at.vector("x")
y_at = at.vector("y")
op1_np = MyDefaultOutputOp()
q_at = op1_np(x_at, y_at)

x_lv, y_lv = var('x'), var('y')
q_et = etuple(op1_np, x_lv, y_lv)

s = unify(q_et, q_at)
assert s[x_lv] == x_at
assert s[y_lv] == y_at
```

Relations in AeMCMC can be written using `nth` and unification will work (we can also special-case the logic for `RandomVariable`s, but we already established we don't want to do that). But to me that seems like a workaround for a much deeper design issue. This PR has been tightly coupled to the design of `RandomVariable`s, which made us use `default_output` for convenience, since the beginning and I need to spend more time thinking about this point.

rlouf commented:
I have fixed the unification with the default outputs of a multiple-output `Apply` nodes. I am now trying to define the behavior of `etuplize` when passed a list of outputs which can happen with the output of multiple output `Op`\s:

```python
from aesara.graph.basic import Apply
from aesara.graph.op import Op
from aesara.graph.type import Type

class MyType(Type):
    def filter(self, data):
        return data

    def __eq__(self, other):
        return isinstance(other, MyType)

    def __hash__(self):
        return hash(MyType)

    def __repr__(self):
        return "MyType()"

class MyMultiOutOp(Op):
    def make_node(self, *inputs):
        outputs = [MyType()(), MyType()()]
        return Apply(self, list(inputs), outputs)

    def perform(self, node, inputs, outputs):
        outputs[0] = np.array(inputs[0])
        outputs[1] = np.array(inputs[0])
```

Etuplizing directly will work, but evaluating the resulting `etuple` will fail:

```python
import aesara.tensor as at
from etuples import etuplize

x_at = at.vector("x")
y_at = at.vector("y")
op1_np = MyMultiOutOp()
z_at = op1_np()

z_et = etuplize(z_at)
z_et._evaled_obj = z_et.null
try:
    z_et.evaled_obj
except Exception as e:
    print(e)
# ExpressionTuple does not have a callable operator.
```

This can be solved by etuplizing lists to `etuple(cons, *args, [])` (which should probably be done in the etuples library directly):

```python
from cons import cons
from cons.core import _car, _cdr
from etuples import etuplize


def car_List(x):
    return cons

_car.add((list,), car_List)

def cdr_List(x):
    return  (*x, [])

_cdr.add((list,), cdr_List)

print(etuplize(z_at))
# e(<class 'cons.core.ConsPair'>, MyMultiOutOp.0, MyMultiOutOp.1, [])

z_et = etuplize(z_at)
z_et._evaled_obj = z_et.null
print(z_et.evaled_obj)
# [MyMultiOutOp.0, MyMultiOutOp.1]
```

However there is still one issue with the previous snippet: `etuplize` did not work recursively. That's because after etuplizing the list to `e(cons, *args, [])` the etuplizing function evaluates the `isinstance(x, ConsPair)` branch which returns the etuple directly.

brandonwillard reviewed (commented):
On a related note: this is reminding me that we **really** need hash-consing and/or caching for these `etuple` conversions.

rlouf commented on aesara/graph/rewriting/unify.py:
```diff
@@ -72,19 +73,90 @@ def __repr__(self):
+    return etuple_Op
+
+
+def nth(n: int, node: Apply):
```
Because I forgot. I added it.

rlouf commented:
This PR got a little out of hand, and I will leave etuplizing a list of outputs for another PR as it seems more complicated than I originally thought, and is non-blocking for now. Let's come back to our original problem, which was being able to have "true" relations in AeMCMC, in the sense that they can work both ways, and merge this PR when we are able to do that.

## Summary

The first issue we ran into was that `__call__` of `RandomVariable`s does not have the same signature as `make_node`, which lead to errors during evaluation. This was solved by creating a `OpExpressionTuple` and modifying some of the etuplization and evalution logic in `etuples`.

Then we realized that when we evaluated the expression tuples obtained after etuplizing `RandomVariable`s gave us the two outputs that are generated by `make_node` instead of the just the `Variable`. That's because Aesara doesn't have an operator to say "This `Variable` is the nth output of the `Apply` node". So we faked that behavior by adding a `nth` operator when etuplizing `Variables` that are the output of a multiple-output `Apply` node. _This should be ported to Aesara's IR_.

And then we realized that we could not unify, say `etuple(etuplize(at.random.normal), var(), var(), var(), 0, 1)` with `etuplize(at.random.normal(0, 1))` since the latter starts with `etuple(nth, 1, ...)`. So we modified the unifying logic to be able to do that with `Variable`s that are a default output (we don't do it with multiple-output nodes as in this case `nth` *must* be specified if we want a particular output). _This makes me doubt the conceptual validity of the use of `default_output`
